jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/assignment/impl/strategy/RoundRobinAssignmentStrategy.java

```diff
@@ -0,0 +1,118 @@
+import java.util.stream.Collectors;
+
+public class RoundRobinAssignmentStrategy implements AssignmentStrategy {
+    private static final Logger logger = LoggerFactory.getLogger(RoundRobinAssignmentStrategy.class);
+    private static final String IDENTIFIER = "RoundRobin";
+
+    private boolean active = IDENTIFIER.equals(System.getProperty("org.jbpm.task.assignment.strategy"));
+
+    private Map<String,CircularQueue> circularQueueMap = new ConcurrentHashMap();
+
+    private class CircularQueue<T> extends LinkedBlockingQueue<T> {
+        @Override
+        public T take() {
+            T headValue = null;
+            try {
+                headValue = super.take();
+                super.offer(headValue);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+            return headValue;
+        }
+    }
+
+    @Override
+    public String getIdentifier() {
+        return IDENTIFIER;
+    }
+
+    @Override
+    public Assignment apply(Task task, TaskContext taskContext, String s) {
+        List<OrganizationalEntity> excluded = ((InternalPeopleAssignments)task.getPeopleAssignments()).getExcludedOwners();
+        UserInfo userInfo = (UserInfo) ((org.jbpm.services.task.commands.TaskContext)taskContext).get(EnvironmentName.TASK_USER_INFO);
+
+        // Get the the users from the task's the potential owners
+        List<OrganizationalEntity> potentialOwners = task.getPeopleAssignments().getPotentialOwners().parallelStream()
+                .filter(oe -> oe instanceof User && !excluded.contains(oe))
+                .collect(Collectors.toList());
+
+        // Get the users belonging to groups that are potential owners
+        task.getPeopleAssignments().getPotentialOwners().parallelStream().filter(oe -> oe instanceof Group)
+                .forEach(oe -> {
+                    Iterator<OrganizationalEntity> groupUsers = userInfo.getMembersForGroup((Group)oe);
+                    if (groupUsers != null) {
+                        groupUsers.forEachRemaining(user -> {
+                            if (user != null && !excluded.contains(user) && !potentialOwners.contains(user)) {
+                                potentialOwners.add(user);
+                            }
+                        });
+                    }
+                });
+        String queueName = getQueueName(task);
+        CircularQueue<OrganizationalEntity> mappedQueue = null;
+        // If a queue already exists for this task then its contents should be synchronized with the
+        // current list of potential owners
+        if (circularQueueMap.containsKey(queueName)) {
+            final CircularQueue<OrganizationalEntity> queue = circularQueueMap.get(queueName); 
+            potentialOwners.forEach(po -> {
+            	if (!queueContainsUser(queue,po)) {
+            		queue.add(po);
+            	}
+            });
+            queue.removeIf(oe -> !potentialOwners.contains(oe));
```

Comment:
Are the above queue modifications thread safe? Asking as this will be executed by multiple threads in parallel as the strategy is shared for all incoming requests. What would happen when there are two concurrent requests fir the same task but they have different potential owner lists?